### ISSUE TITLE
Add a StreamTransformer constructor that takes a bind callback

### ISSUE THREAD

**Issue by nex3** [dart-lang/sdk#32021]
It currently requires a lot of overhead to write a `StreamTransformer` that takes a `Stream` as input and returns a different `Stream`. For example, if I want to create a transformer that parses a stream of JSON into objects and filters out non-Maps, I have to either create and instantiate an entire subclass of `StreamTransformerBase` or call `new StreamTransformer()` and manually forward `Stream.listen()`. I'd much prefer to be able to write:

```dart
final jsonMapTransformer = new StreamTransformer.fromCallback((stream) => json
    .decoder
    .bind(stream)
    .where((object) => object is Map));
```

**Comment by lrhn** [comment]
Agree. The `StreamTransformer` API is really just the `bind` function, so creating a transformer form a function implementing that method would be nice.
(Since we don't have Java's anonymous classes).
